added bfs solution for milkpails (#681)
cpinitiative/usaco-guide: solutions/USACO feb16-milk-pails.mdx

id: usaco-620
source: USACO Silver 2016 February
title: Milk Pails
author: Óscar Garries, Neo Wang
---

[Official Analysis](http://www.usaco.org/current/data/sol_pails_silver_feb16.html)

<LanguageSection>
<CPPSection>

> ## C++ Implementation (BFS)

Reviewer: Can you make this an alternative sol because technically bfs isn't in silver?

const int MOD = 1e9 + 7;

int dist[MX][MX];

void setIn(string s) { freopen(s.c_str(),"r",stdin); }
void setOut(string s) { freopen(s.c_str(),"w",stdout); }

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    setIn("pails.in");
    setOut("pails.out");

    cin >> X >> Y >> K >> M;

    F0R(i, MX) F0R(j, MX) dist[i][j] = MOD;

    queue<pair<int, int>> bfs;
    bfs.push({0, 0});
    dist[0][0] = 0;

    while(!bfs.empty()) {
        pair<int, int> top = bfs.front();
        bfs.pop();
        int ndist = dist[top.first][top.second]+1;

        int pourX = min(top.first, Y - top.second);
        int pourY = min(top.second, X - top.first);

        int nx[] = {top.first, 0, X, top.first, top.first - pourX, top.first + pourY }; // all possible pours
        int ny[] = {0, top.second, top.second, Y, top.second + pourX, top.second - pourY };

        F0R(i, 6) {
            if(ndist > K || dist[nx[i]][ny[i]] != MOD) continue;
            dist[nx[i]][ny[i]] = ndist;
            bfs.push({nx[i], ny[i]});
        }
    }

    int ret = MOD;

    F0R(i, MX) {
        F0R(j, MX) {
            if(dist[i][j] != MOD) ret = min(ret, abs(M - (i+j)));
        }
    }

>     cout << ret;

Author: hm thats strange i haven't done endl and it has always worked for me but i don't see why not

id: usaco-620
source: USACO Silver 2016 February
title: Milk Pails
author: Óscar Garries, Neo Wang
---

[Official Analysis](http://www.usaco.org/current/data/sol_pails_silver_feb16.html)

<LanguageSection>
<CPPSection>

## C++ Implementation (BFS)

We can directly simulate all operations using BFS.

> In this problem, we only care about states $(i, j)$ where pail $X$ has $i$ and pail $Y$ has $j$ units of milk. We can perform a BFS starting from base state $(0, 0)$ to compute $\texttt{dist}[i][j]$, the minimum number of steps to reach each state. Afterwards, compute the answer by iterating over all states where $\texttt{dist}[i][j] \le K$ and finding the minimum $|i+j-M|$.

Reviewer 2: ```suggestion
In this problem, we care about states $(i, j)$ where pail $X$ has $i$ and pail $Y$ has $j$ units of milk. We can perform a BFS starting from base state $(0, 0)$ to compute $\texttt{dist}[i][j]$, the minimum number of steps to reach each state. Afterwards, compute the answer by iterating over all states where $\texttt{dist}[i][j] \le K$ and finding the minimum $|i+j-M|$.
```
(yea my original wording wasn't perfect, and it still isn't, but since there's another comment anyways I might as well add some small edits to make it a bit better)

For every transition in the BFS, we simulate all possible actions.

1. Fill either pail completely to the top: set $i$ to $X$ or $j$ to $Y$.
2. Empty either pail: set $i$ or $j$ to $0$.
3. Pour the contents of one pail into another without overflowing or running out of milk.

> To perform operation 3, we can find the amount poured to be $\min(i, X-j)$ or $\min(j, Y-i)$ depending on which pail you pour from. Then add/subtract this quantity to each pail appropriately.

Reviewer 2: ```suggestion
To perform operation 3, we can find the amount poured to be $\min(i, Y-j)$ or $\min(j, X-i)$ depending on which pail you pour from. Then add/subtract this quantity to each pail appropriately.
```

You accidentally reverted to the typo :P

Now iterate over all states $(i, j)$ where $\texttt{dist}[i][j] \le K$ and compute the minimum $|i+j-M|$.

```cpp
#include <bits/stdc++.h>

using namespace std;

#define FOR(i, a, b) for(int i = (a); i < (b); i++)
#define FORE(i, a, b) for(int i = (a); i <= (b); i++)
#define F0R(i, a) for(int i = 0; i < (a); i++)
#define trav(a, x) for (auto& a : x)

int X, Y, K, M;

const int MX = 101;

> const int MOD = 1e9 + 7;

Reviewer 3: i think you should change this to INF to make it less confusing

id: usaco-620
source: USACO Silver 2016 February
title: Milk Pails
author: Óscar Garries, Neo Wang
---

[Official Analysis](http://www.usaco.org/current/data/sol_pails_silver_feb16.html)

<LanguageSection>
<CPPSection>

## C++ Implementation (BFS)

We can directly simulate all operations using BFS.

> When we are performing our BFS, we only have to keep track of the amount of operations used so far for the current amount of milk in pails $X$ and $Y$. That way, we can check all nodes with distance $\le K$ once our BFS is finished, and then compute the smallest value of  $|i + j -M|$.

Reviewer 2: ```suggestion
In this problem, we only care about states $(i, j)$ where pail $X$ has $i$ and pail $Y$ has $j$ units of milk. We can perform a BFS starting from base state $(0, 0)$ to compute $\texttt{dist}[i][j]$, the minimum number of steps to reach each state. Afterwards, compute the answer by iterating over all states where $\texttt{dist}[i][j] \le K$ and finding the minimum $|i+j-M|$.
```

> In our matrix, the current amount of milk in $X$ is $i$, and $Y$ is $j$.

Reviewer 2: This line is not necessary anymore

> For every step, we simulate all the possible actions.

Reviewer 2: ```suggestion
For every transition in the BFS, we simulate all possible actions.
```

1. Fill either pail completely to the top: set $i$ to $X$ or $j$ to $Y$.
2. Empty either pail: set $i$ or $j$ to $0$.
3. Pour the contents of one pail into another without overflowing or running out of milk.

> To perform operation 3, let the quantity to be poured $Q$ be denoted by either $Q_x=\min(i, X - j)$ or $Q_y=\min(j, Y-i)$, depending on which bucket is poured ($Q_x$ if we are pouring out of $x$ and vice versa). Then, to pour $X$ into $Y$, we subtract this poured quantity $Q_x$ from $X$ and add it to $Y$. In other words, if we were to pour $X$ into $Y$, the new amount in $X$ would be $X - Q_x$, and the new amount in $Y$ is $Y + Q_x$. To pour $Y$ into $X$, this would be $Y_{\textrm{new}}=Y-Q_y$, $X_{\textrm{new}}=X+Q_y$.

Reviewer 3: ```suggestion
To perform operation 3, let the quantity to be poured $Q$ be denoted by either $Q_x=\min(i, Y - j)$ or $Q_y=\min(j, X-i)$, depending on which bucket is poured ($Q_x$ if we are pouring out of $x$ and vice versa). Then, to pour $X$ into $Y$, we subtract this poured quantity $Q_x$ from $X$ and add it to $Y$. In other words, if we were to pour $X$ into $Y$, the new amount in $X$ would be $X - Q_x$, and the new amount in $Y$ is $Y + Q_x$. To pour $Y$ into $X$, this would be $Y_{\textrm{new}}=Y-Q_y$, $X_{\textrm{new}}=X+Q_y$.
```

Reviewer 3: typo?

Author: oops

Reviewer 2: Nice catch!

> To perform operation 3, let the quantity to be poured $Q$ be denoted by either $Q_x=\min(i, Y - j)$ or $Q_y=\min(j, X-i)$, depending on which bucket is poured ($Q_x$ if we are pouring out of $x$ and vice versa). Then, to pour $X$ into $Y$, we subtract this poured quantity $Q_x$ from $X$ and add it to $Y$. In other words, if we were to pour $X$ into $Y$, the new amount in $X$ would be $X - Q_x$, and the new amount in $Y$ is $Y + Q_x$. To pour $Y$ into $X$, this would be $Y_{\textrm{new}}=Y-Q_y$, $X_{\textrm{new}}=X+Q_y$.

Reviewer 2: I think this is a bit excessive. Isn't what you had earlier ok? (except the typo)

> Now, to compute the smallest $|i+j-M|$, simply scan over all elements where $A[i][j] \le K$.

Reviewer 2: ```suggestion
Now iterate over all states $(i, j)$ where $\texttt{dist}[i][j] \le K$ and compute the minimum $|i+j-M|$.
```

Double check latex & formatting

#define trav(a, x) for (auto& a : x)

int X, Y, K, M;

const int MX = 101;
const int MOD = 1e9 + 7;

int dist[MX][MX];

void setIn(string s) { freopen(s.c_str(),"r",stdin); }
void setOut(string s) { freopen(s.c_str(),"w",stdout); }

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    setIn("pails.in");
    setOut("pails.out");

    cin >> X >> Y >> K >> M;

    F0R(i, MX) F0R(j, MX) dist[i][j] = MOD;

    queue<pair<int, int>> bfs;
    bfs.push({0, 0});
    dist[0][0] = 0;

    while(!bfs.empty()) {
        pair<int, int> top = bfs.front();
        bfs.pop();
        int ndist = dist[top.first][top.second]+1;

        int pourX = min(top.first, Y - top.second);
        int pourY = min(top.second, X - top.first);

        int nx[] = {top.first, 0, X, top.first, top.first - pourX, top.first + pourY }; // all possible pours
        int ny[] = {0, top.second, top.second, Y, top.second + pourX, top.second - pourY };

        F0R(i, 6) {
            if(ndist > K || dist[nx[i]][ny[i]] != MOD) continue;
            dist[nx[i]][ny[i]] = ndist;
            bfs.push({nx[i], ny[i]});
        }
    }

    int ret = MOD;

    F0R(i, MX) {
        F0R(j, MX) {

>             if(dist[i][j] != MOD) ret = min(ret, abs(M - (i+j)));

Reviewer 2: For consistency with explanation:

```suggestion
            if(dist[i][j] <= K) ret = min(ret, abs(M - (i+j)));
```
